Preserve microphone and camera permissions in case of same-origin web application initiated navigations

youennf opened the pull request:
#### a8ed1887e6b234660829c3d6f8978149b276a1b5
<pre>
Preserve microphone and camera permissions in case of same-origin web application initiated navigations
<a href="https://rdar.apple.com/138122655">rdar://138122655</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282339">https://bugs.webkit.org/show_bug.cgi?id=282339</a>

Reviewed by Eric Carlson.

Before the patch, camnera/microphone permissions are reset for every navigation.

When a web application navigates directly without the user involvement, the user may not notice the navigation.
And user may be annoyed by being asked again for microphone/camera prompts.

We change the heuristic to keep resetting camnera/microphone permissions in case:
- User decided for the navigation (whether sameo origin or not) or the reload
- Navigation is cross-origin.
Othewrwise, we no longer reset these permissions.

To do so, within WebPageProxy::didChangeMainDocument, we check whether the navigation is user initiated or not.
We fix a case where reload triggered by a user was not always marked as user initiated by updating WebPage::reload and FrameLoader::reload.

We make WebLocalFrameLoaderClient::dispatchDidChangeMainDocument pass the navigation ID so that WebPageProxy::didChangeMainDocument can get the navigation information.

Covered by API test.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChangeMainDocument):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidChangeMainDocument):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::reload):
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
(TestWebKitAPI::(WebKit, KeepPermissionForWebAppSameOriginNavigations)):
(TestWebKitAPI::InterruptionBetweenGetDisplayMediaAndGetUserMedia)): Deleted.

Canonical link: <a href="https://commits.webkit.org/285950@main">https://commits.webkit.org/285950@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad9a845d0f019da0b33838712aa1a56dc8efed5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74199 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27010 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78573 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25437 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76316 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1413 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58325 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16670 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48489 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38735 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45417 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21320 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23770 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66875 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80092 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/853 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66630 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1661 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63841 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65908 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16369 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9838 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7996 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1480 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4268 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1509 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1497 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1516 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2281,7 +2281,8 @@ void WebPage::reload(WebCore::NavigationIdentifier navigationID, OptionSet<WebCo
     m_sandboxExtensionTracker.beginReload(mainFrame.ptr(), WTFMove(sandboxExtensionHandle));
     if (m_page && mainFrame->coreLocalFrame()) {
-        mainFrame->coreLocalFrame()->loader().reload(reloadOptions);
+        bool isRequestFromClientOrUserInput = true;
+        mainFrame->coreLocalFrame()->loader().reload(reloadOptions, isRequestFromClientOrUserInput);
```
Nit: is the local necessary?

youennf replied:
It serves as a documentation to make it clear what the bool is.

youennf committed:
Preserve microphone and camera permissions in case of same-origin web application initiated navigations
rdar://138122655
https://bugs.webkit.org/show_bug.cgi?id=282339

Reviewed by Eric Carlson.

Before the patch, camnera/microphone permissions are reset for every navigation.

When a web application navigates directly without the user involvement, the user may not notice the navigation.
And user may be annoyed by being asked again for microphone/camera prompts.

We change the heuristic to keep resetting camnera/microphone permissions in case:
- User decided for the navigation (whether sameo origin or not) or the reload
- Navigation is cross-origin.
Othewrwise, we no longer reset these permissions.

To do so, within WebPageProxy::didChangeMainDocument, we check whether the navigation is user initiated or not.
We fix a case where reload triggered by a user was not always marked as user initiated by updating WebPage::reload and FrameLoader::reload.

We make WebLocalFrameLoaderClient::dispatchDidChangeMainDocument pass the navigation ID so that WebPageProxy::didChangeMainDocument can get the navigation information.

Covered by API test.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::resetAccess):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChangeMainDocument):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidChangeMainDocument):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::reload):
* Tools/TestWebKitAPI/Tests/WebKit/GetUserMedia.mm:
(TestWebKitAPI::(WebKit, KeepPermissionForWebAppSameOriginNavigations)):
(TestWebKitAPI::InterruptionBetweenGetDisplayMediaAndGetUserMedia)): Deleted.

Canonical link: https://commits.webkit.org/285950@main
